lib/schema/basal.js

```diff
@@ -100,39 +156,48 @@ module.exports = function(streamDAO){
           if (annotate) {
             var adjustedPrev = adjustDuration(datum, disjointPrev);
             if (adjustedPrev != null) {
-              eventsToStore.push(schema.annotateEvent(
+              // annotation makes a clone
+              const annotated = schema.annotateEvent(
                 adjustedPrev,
                 { code: mismatchedSeries, nextId: schema.generateId(datum, idFields) }
-              ));
+              );
+              updateDatum(annotated);
             }
           }
         }
-        done();
-      });
-    }
-
+      }
 
-    if (datum.previous != null) {
-      var prevId = schema.generateId(datum.previous, idFields);
+      array.forEach(datum => streamDAO.ensureInternalId(datum));
 
-      streamDAO.getDatum(prevId, datum._groupId, function(err, actualPrev){
-        if (err != null) {
-          return cb(err);
-        }
+      // Handle the input array. The input array may modify data not inside
+      // the input array such as previous entries in the db leading to more
+      // data being inserted/modified than the size of the original input
+      // array, dataOrDatum.
+      for (let datum of array) {
+        if (datum.previous != null) {
+          var prevId = schema.generateId(datum.previous, idFields);
+          var prevInternalId = streamDAO.generateInternalId(prevId, datum._groupId);
+          const actualPrev = latestData[prevInternalId];
 
-        if (actualPrev != null) {
-          var adjustedPrev = adjustDuration(datum, actualPrev);
-          if (adjustedPrev != null && !_.isEqual(adjustedPrev, actualPrev)) {
-            eventsToStore.push(adjustedPrev);
+          if (actualPrev != null) {
+            var adjustedPrev = adjustDuration(datum, actualPrev);
+            if (adjustedPrev != null && !_.isEqual(adjustedPrev, actualPrev)) {
+              updateDatum(adjustedPrev);
+            }
+          } else {
+            annotateDatumBefore(datum);
```

Comment:
I'm a bit confused here. The above code seems to determine the previous datum (either from `datum.previous` or finding the `actualPrev` by id generation. Why then are we using `annotateDatumBefore` (which searches by the time field via `getLatestDatumBefore`)? Don't we already have the previous datum that needs to be updated? Why the second search for previous by time?

I must be missing something.

Reply:
The logic is based on the previous code and worked like so:
```js
if (datum_previous_is_set_and_exists) {
  annotate_previou();
} else {
  annotate_datum_that_has_latest_time_before_current_datums_time()
}
```